apps/explorer/lib/explorer/smart_contract/helper.ex

```diff
@@ -53,4 +57,56 @@ defmodule Explorer.SmartContract.Helper do
 
     "0x" <> topic
   end
+
+  @doc "Return all events on contract, return also implementation events if contract is a proxy."
+  def get_all_events(%SmartContract{address_hash: address, abi: abi} = contract) do
+    proxy = Chain.proxy_contract?(address, abi)
+
+    events =
+      if proxy do
+        implementation_events =
+          contract
+          |> get_implementation_contract()
+          |> then(&filter_events(&1.abi))
+
+        implementation_events ++ filter_events(contract.abi)
+      else
+        filter_events(contract.abi)
+      end
+
+    # add the topic directly on the abi (not actually part of the abi itself but used ubiquitously)
+    # then dedup by the topic
+    events
+    |> Enum.map(fn event -> Map.put(event, "topic", event_abi_to_topic_str(event)) end)
+    |> Enum.uniq_by(&Map.get(&1, "topic"))
+  end
+
+  defp get_implementation_contract(%SmartContract{address_hash: address_hash, abi: abi}) do
```

Comment:
Can't we rely on `Chain.get_implementation_abi_from_proxy` method?

Reply:
we could, but it has some caveats including 

1) it will return the[ abi of a contract with matching bytecode](https://github.com/celo-org/blockscout/blob/2f99f93c3585c27913beaa852c3b73421ed3ef98/apps/explorer/lib/explorer/chain.ex#L4621) should the implementation address not be verified
     * which may be desirable as it is indeed the same bytecode, but I wanted to assert that the actual implementation was verified before continuing
2) as a failure mode [it will return an empty abi](https://github.com/celo-org/blockscout/blob/2f99f93c3585c27913beaa852c3b73421ed3ef98/apps/explorer/lib/explorer/chain.ex#L7748)
     * again I would rather assert the contents of the implementation itself, if the implementation contract is indeed an empty contract (somehow) then this can be returned from a genuine abi lookup

Reply:
Okay, if the goal is to enforce stricter checks, it makes sense to me.